xcpmd/src/battery.c

```diff
@@ -0,0 +1,803 @@
+
+    //Ensure the directory exists before trying to write the leaves
+    make_xenstore_battery_dir(battery_index);
+
+    //Now write the leaves.
+    sprintf(xenstore_path, "%s%i/%s", XS_BATTERY_PATH, battery_index, XS_BIF_LEAF);
+    xenstore_write(bif, xenstore_path);
+
+
+    //Here for compatibility--will be removed eventually
+    if (battery_index == 0)
+        xenstore_write(bif, XS_BIF);
+    else
+        xenstore_write(bif, XS_BIF1);
+}
+
+
+//Exactly what it says on the tin.
+void write_battery_status_to_xenstore(unsigned int battery_index) {
+
+    struct battery_status * status;
+    char bst[35], xenstore_path[128];
+    int num_batteries, current_battery_level;
+
+    if (battery_index >= num_battery_structs_allocd) {
+        cleanup_removed_battery(battery_index);
+    }
+
+    num_batteries = get_num_batteries_present();
+    if (num_batteries == 0) {
+        xenstore_write("0", XS_BATTERY_PRESENT);
+        return;
+    }
+    else {
+        xenstore_write("1", XS_BATTERY_PRESENT);
+    }
+
+    status = &last_status[battery_index];
+
+    //Delete the BST and reset the "present" flag if the battery is not currently present.
+    if (status->present != YES) {
+
+        sprintf(xenstore_path, "%s%i/%s", XS_BATTERY_PATH, battery_index, XS_BST_LEAF);
+        xenstore_rm(xenstore_path);
+
+        sprintf(xenstore_path, "%s%i/%s", XS_BATTERY_PATH, battery_index, XS_BATTERY_PRESENT_LEAF);
+        xenstore_write("0", xenstore_path);
+        return;
+    }
+
+    //Build the BST structure.
+    memset(bst, 0, 35);
+    snprintf(bst, 3, "%02x", 16);
+    write_ulong_lsb_first(bst+2, status->state);
+    write_ulong_lsb_first(bst+10, status->present_rate);
+    write_ulong_lsb_first(bst+18, status->remaining_capacity);
+    write_ulong_lsb_first(bst+26, status->present_voltage);
+
+    //Ensure the directory exists before trying to write the leaves
+    make_xenstore_battery_dir(battery_index);
+
+    //Now write the leaves.
+    sprintf(xenstore_path, XS_BATTERY_PATH "%i/" XS_BST_LEAF, battery_index);
+    xenstore_write(bst, xenstore_path);
+
+    sprintf(xenstore_path, "%s%i/%s", XS_BATTERY_PATH, battery_index, XS_BATTERY_PRESENT_LEAF);
+    xenstore_write("1", xenstore_path);
+
+    //Here for compatibility--will be removed eventually
+    if (battery_index == 0)
+        xenstore_write(bst, XS_BST);
+    else
+        xenstore_write(bst, XS_BST1);
+
+    current_battery_level = get_current_battery_level();
+    if (current_battery_level == NORMAL || get_ac_adapter_status() == ON_AC)
+        xenstore_rm(XS_CURRENT_BATTERY_LEVEL);
+    else {
+        xenstore_write_int(current_battery_level, XS_CURRENT_BATTERY_LEVEL);
+        notify_com_citrix_xenclient_xcpmd_battery_level_notification(xcdbus_conn, XCPMD_SERVICE, XCPMD_PATH);
+        xcpmd_log(LOG_ALERT, "Battery level below normal - %d!\n", current_battery_level);
+    }
+
+#ifdef XCPMD_DEBUG
+    //xcpmd_log(LOG_DEBUG, "~Updated battery information in xenstore\n");
```

Comment:
Another bit I commented out during my own debugging... I'll uncomment it for everyone else.